lib/include/atomic.h

```diff
@@ -0,0 +1,484 @@
+ * and then setting/clearing configUSE_ATOMIC_INSTRUCTION in FreeRTOSConfig.h.
+ *
+ * Define AND set configUSE_ATOMIC_INSTRUCTION to 1 for ISA native atomic support.
+ * Undefine OR clear configUSE_ATOMIC_INSTRUCTION for disabling global interrupt implementation.
+ */
+
+#ifndef ATOMIC_H
+#define ATOMIC_H
+
+#ifndef INC_FREERTOS_H
+    #error "include FreeRTOS.h must appear in source files before include atomic.h"
+#endif
+
+#ifndef PORTMACRO_H
+    #error "include portmacro.h must appear in source files before include atomic.h"
+#endif
+
+/* Standard includes. */
+#include <stdint.h>
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined ( configUSE_ATOMIC_INSTRUCTION ) && ( configUSE_ATOMIC_INSTRUCTION == 1 )
+
+    /* This branch is for GCC compiler and GCC compiler only. */
+    #ifndef portFORCE_INLINE
+        #define portFORCE_INLINE  inline __attribute__((always_inline))
+    #endif
+
+#else
+
+    /* Port specific definitions --
+     * how to enter/exit critical section from atomic.
+     * Refer template -- ./lib/FreeRTOS/portable/Compiler/Arch/portmacro.h
+     */
+    #if defined( portSET_INTERRUPT_MASK_FROM_ISR )
+        /* Nested interrupt scheme is supported in this port. */
+        #define ATOMIC_ENTER_CRITICAL()     portSET_INTERRUPT_MASK_FROM_ISR()
+        #define ATOMIC_EXIT_CRITICAL( x )   portCLEAR_INTERRUPT_MASK_FROM_ISR( x )
+        typedef UBaseType_t CriticalSessionType_t;
+    #else
+        #define ATOMIC_ENTER_CRITICAL()     portENTER_CRITICAL()
+        #define ATOMIC_EXIT_CRITICAL( x )   portEXIT_CRITICAL()
+        typedef void CriticalSessionType_t;
+    #endif /* portSET_INTERRUPT_MASK_FROM_ISR() */
+
+    /* Port specific definitions --
+     * how to ensure "always inline".
+     * portFORCE_INLINE is used. If this is not defined, add the definition to the corresponding portmacro.h.
+     */
+
+#endif /* configUSE_ATOMIC_INSTRUCTION */
+
+/*----------------------------- Swap && CAS ------------------------------*/
+
+/**
+ * Atomic compare-and-swap
+ *
+ * @brief Performs an atomic compare-and-swap operation on the specified values.
+ *
+ * @param[in, out] pDestination  Pointer to memory location from where value is to be loaded and checked.
+ * @param[in] ulExchange         If condition meets, write this value to memory.
+ * @param[in] ulComparand        Swap condition, checks and waits for *pDestination to be equal to ulComparand.
+ *
+ * @return bool True for swapped, false for not swapped.
+ *
+ * @note This function only swaps *pDestination with ulExchange, if previous *pDestination value equals ulComparand.
+ */
+static portFORCE_INLINE bool Atomic_CompareAndSwap_u32( uint32_t volatile * pDestination, uint32_t ulExchange, uint32_t ulComparand )
+{
+
+#if defined ( configUSE_ATOMIC_INSTRUCTION ) && ( configUSE_ATOMIC_INSTRUCTION == 1 )
+    return __atomic_compare_exchange( pDestination, &ulComparand, &ulExchange, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST );
```

Comment:
Should w add a few words (and/or links) here to explain what the flags __ATOMIC_xxx are?

Reply:
link to gcc is given in comment section.